radix-engine/src/track/track.rs

```diff
@@ -17,50 +17,97 @@ use sbor::rust::mem;
 
 use super::interface::{CanonicalPartition, CanonicalSubstateKey, StoreCommit, StoreCommitInfo};
 
+/// A captured sequence of [`StateUpdate`]s to be applied in particular order.
 #[derive(Debug, Clone, PartialEq, Eq, ScryptoSbor, Default)]
 pub struct StateUpdates {
-    /// A set of partitions that were entirely deleted.
-    /// Note: this batch changes should be applied *before* the [`system_updates`] below (i.e.
-    /// allowing a latter individual substate creation to be applied).
-    pub partition_deletions: IndexSet<(NodeId, PartitionNumber)>,
-
-    /// A set of individual substate updates (indexed by partition and substate key).
-    pub system_updates: SystemUpdates,
+    pub elements: Vec<StateUpdate>,
 }
 
 impl StateUpdates {
-    /// Merges the updates from the given instance into this one.
-    pub fn extend(&mut self, mut other: Self) {
-        self.partition_deletions
-            .extend(other.partition_deletions.drain(..));
-        self.system_updates.extend(other.system_updates.drain(..));
+    /// Creates an instance from the previously used representation, consisting of:
+    /// - partitions to be entirely deleted (applied first),
+    /// - individual substate changes, indexed by partition and substate key.
+    pub fn from_legacy_representation(
+        partition_deletions: IndexSet<(NodeId, PartitionNumber)>,
+        system_updates: IndexMap<(NodeId, PartitionNumber), IndexMap<SubstateKey, DatabaseUpdate>>,
+    ) -> Self {
+        Self {
+            elements: partition_deletions
+                .into_iter()
+                .map(|(node_id, partition_num)| {
+                    StateUpdate::Batch(BatchStateUpdate::DeletePartition(node_id, partition_num))
+                })
+                .chain(system_updates.into_iter().flat_map(
+                    |((node_id, partition_num), by_substate_key)| {
+                        by_substate_key
+                            .into_iter()
+                            .map(move |(substate_key, update)| {
+                                StateUpdate::Single(SingleSubstateUpdate {
+                                    node_id: node_id.clone(),
```

Comment:
_(note from author)_

This line shows one implication:
If a million substates from the same Node is changed, than this new `Vec<>` representation wastes 33MB on repeating the same `NodeId` million times.
If this is a problem, then we can migrate from `Vec<StateUpdate>` to `IndexMap<NodeId, PartitionStateUpdate>` (I assume the order of applying changes between Nodes will _never_ matter?).
_But my default vote is "do not drive API design by performance aspects before performance problems occur or become obvious"._

Reply:
I think that logic makes sense (I wouldn't consider 33MB to be THAT bad). I am a little more skeptical though about the lost information regarding the shared node/partition updates. It looks like we re-index these when converting to `DatabaseUpdates`. Is there a reason to remove the index?

Reply:
> It looks like we re-index these when converting to `DatabaseUpdates`.

Yes, this is true at the moment: we lose the indexing effort and need to re-index into basically-the-same-structure-that-Track-already-had == `DatabaseUpdates`.

However, we are now free to turn the `DatabaseUpdates` into a "canonical change" (i.e. something like `TreeChange` that I needed at https://github.com/radixdlt/radixdlt-scrypto/pull/1430/files#diff-8081b9309cd2a88224cb5012df79c25edd11493cc545b59c3fb772db191a1c0fR11). I actually want to do it as part of "support partition delete in all storages". 
And this will need slightly different indexing anyways (i.e. a canonicalization process, which does `Vec<ChangeOp> -> TreeChange`). So I effectively do not need that previous indexing.

Reply:
I guess the indexing doesn't work any more due to the partition deletes.

But I still wonder if we can actually encode the canonicalization into the form, and still get to keep the indexing, eg, as:
```
StateUpdates {
   initial_partition_deletions: IndexMap<PartitionDeletes>,
   substate_changes: IndexMap<(..), SubstateChange>,
}
```
Which I guess is kinda what we started with :')

Reply:
I'm not worried about the state size, but I do think we'll end up recreating that index (key) => change in a lot of places if we don't have it yet.

Reply:
The https://github.com/radixdlt/radixdlt-scrypto/pull/1454#discussion_r1315712082 suggests that we don't want to rush and merge this mid-step into rcnet-v3.1 - and I kinda agree.

We can schedule a slower migration towards a true canonical form (which can also include indexing, if we find it useful!). Something like:
```
struct StateUpdates {
  by_node: IndexMap<NodeId, NodeTierUpdates>,
}

enum NodeTierUpdates {
  Delta { by_partition: IndexMap<PartitionNumber, PartitionTierUpdates> },
  // possible place to add Batch(DeleteNode) or whatever we need later
}

enum PartitionTierUpdates {
  Delta { by_substate: IndexMap<SubstateKey, SubstateUpdate> }, // individual creations/updates/deletes
  Batch(BatchPartitionUpdate),
}

enum BatchPartitionUpdate {
  Reset { new_contents: IndexMap<SubstateKey, SubstateValue> }, // note: Reset >= Delete
}
```

Reply:
The proposed canonical form loses details about the timeline-order of substate updates inside a transaction. Not respecting timeline-ordering in our API responses is something that integrators kept nagging us about because they found it unexpected.

They mostly care inside a Node/Partition though, so this might be okay.

But perhaps it's more straightforward to create the following from the canonical StateUpdates ourselves later, and leave the transaction receipt almost as-is?
```rust
struct MergableStateUpdates {
  by_node: IndexMap<NodeId, NodeTierUpdates>,
}
```

Reply:
Oh wait I'm sorry looks like we already don't have that:
```
pub type SystemUpdates = IndexMap<(NodeId, PartitionNumber), IndexMap<SubstateKey, DatabaseUpdate>>;
```

So perhaps it's okay to go a step further and just capture it by Node, Partition, and then Substate in the partition.